### PR TITLE
Fixed reconciliation of offloading patch for calico

### LINKED ISSUE #3695: MicroK8s - Pods with hostNetwork do not have access to cluster DNS
#### Summary
MicroK8s - Pods with hostNetwork do not have access to cluster DNS

#### What Should Happen Instead?

Pods with hostNetwork should have access to cluster DNS and should get resolved. 
I have even tried with dns policy as - ClusterFirstWithHostNet but still couldn't be able to fix this issue. please find the below details.

#### Reproduction Steps

Create busybox pod with hostnetwork and dns policy as ClusterFirstWithHostNet. 
Create another busy box pod without hostnetwork. 
Create nginx pod and expose it with svc. 

nslookup for my nginx pod fails from hostnetwork pod where it gets resolved as expected from the normal one.

root@master:~# kubectl exec -it busybox -- sh
/ # nslookup nginx
Server: 10.152.183.10
Address 1: 10.152.183.10 kube-dns.kube-system.svc.cluster.local

Name: nginx
Address 1: 10.152.183.139 nginx.default.svc.cluster.local
/ # exit
root@master:~#

root@master:~# kubectl exec -it busybox-hostnet -- sh
/ # nslookup nginx
Server: 10.152.183.10
Address 1: 10.152.183.10

nslookup: can't resolve 'nginx'
/ # exit
command terminated with exit code 1

Pod yamls as below:

apiVersion: v1
kind: Pod
metadata:
name: busybox-hostnet
namespace: default
spec:
containers:

image: busybox:1.28
command:
sleep
"3600"
imagePullPolicy: IfNotPresent
name: busybox
restartPolicy: Always
hostNetwork: true
dnsPolicy: ClusterFirstWithHostNet

coredns yamls:

apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"addonmanager.kubernetes.io/mode":"Reconcile","k8s-app":"kube-dns","kubernetes.io/cluster-service":"true","kubernetes.io/name":"CoreDNS"},"name":"coredns","namespace":"kube-system"},"spec":{"selector":{"matchLabels":{"k8s-app":"kube-dns"}},"strategy":{"rollingUpdate":{"maxSurge":"10%","maxUnavailable":0},"type":"RollingUpdate"},"template":{"metadata":{"annotations":{"priorityClassName":"system-cluster-critical"},"labels":{"k8s-app":"kube-dns"}},"spec":{"containers":[{"args":["-conf","/etc/coredns/Corefile"],"image":"coredns/coredns:1.9.3","imagePullPolicy":"IfNotPresent","livenessProbe":{"failureThreshold":5,"httpGet":{"path":"/health","port":8080,"scheme":"HTTP"},"initialDelaySeconds":60,"successThreshold":1,"timeoutSeconds":5},"name":"coredns","ports":[{"containerPort":53,"name":"dns","protocol":"UDP"},{"containerPort":53,"name":"dns-tcp","protocol":"TCP"},{"containerPort":9153,"name":"metrics","protocol":"TCP"}],"readinessProbe":{"httpGet":{"path":"/ready","port":8181,"scheme":"HTTP"}},"resources":{"limits":{"memory":"170Mi"},"requests":{"cpu":"100m","memory":"70Mi"}},"securityContext":{"allowPrivilegeEscalation":false,"capabilities":{"add":["NET_BIND_SERVICE"],"drop":["all"]},"readOnlyRootFilesystem":true},"volumeMounts":[{"mountPath":"/etc/coredns","name":"config-volume","readOnly":true}]}],"dnsPolicy":"Default","priorityClassName":"system-cluster-critical","serviceAccountName":"coredns","tolerations":[{"key":"CriticalAddonsOnly","operator":"Exists"}],"volumes":[{"configMap":{"items":[{"key":"Corefile","path":"Corefile"}],"name":"coredns"},"name":"config-volume"}]}}}}
  creationTimestamp: "2023-01-25T08:44:54Z"
  generation: 7
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    k8s-app: kube-dns
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: CoreDNS
  name: coredns
  namespace: kube-system
  resourceVersion: "39003"
  uid: c6c9c711-1d6a-44ab-a2f7-607406b9a22e
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: kube-dns
  strategy:
    rollingUpdate:
      maxSurge: 10%
      maxUnavailable: 0
    type: RollingUpdate
  template:
    metadata:
      annotations:
        priorityClassName: system-cluster-critical
      creationTimestamp: null
      labels:
        k8s-app: kube-dns
    spec:
      containers:
      - args:
        - -conf
        - /etc/coredns/Corefile
        image: coredns/coredns:1.9.3
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 5
          httpGet:
            path: /health
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: coredns
        ports:
        - containerPort: 53
          name: dns
          protocol: UDP
        - containerPort: 53
          name: dns-tcp
          protocol: TCP
        - containerPort: 9153
          name: metrics
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /ready
            port: 8181
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            memory: 170Mi
          requests:
            cpu: 100m
            memory: 70Mi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            add:
            - NET_BIND_SERVICE
            drop:
            - all
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/coredns
          name: config-volume
          readOnly: true
      dnsPolicy: Default
      priorityClassName: system-cluster-critical
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: coredns
      serviceAccountName: coredns
      terminationGracePeriodSeconds: 30
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      volumes:
      - configMap:
          defaultMode: 420
          items:
          - key: Corefile
            path: Corefile
          name: coredns
        name: config-volume
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2023-01-25T08:44:54Z"
    lastUpdateTime: "2023-01-25T08:44:57Z"
    message: ReplicaSet "coredns-6f5f9b5d74" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2023-01-25T14:56:40Z"
    lastUpdateTime: "2023-01-25T14:56:40Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 7
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1

#### Introspection Report
Tar ball attached. 

[inspection-report-20230125_153600.tar.gz](https://github.com/canonical/microk8s/files/10500871/inspection-report-20230125_153600.tar.gz)

### PR DESCRIPTION
**Pull request by berkayoz**
Closes #3695

#### Summary

Moved reconciliation of offloading patch from the reload process
#### Changes

Moved reconciliation of offloading patch from the reload process and added a flag so it's run on every install/upgrade
#### Testing

Manually tested.

#### Checklist

* [x] Read the [contributions](https://github.com/canonical/microk8s/blob/master/CONTRIBUTING.md) page.
* [x] Submitted the [CLA form](https://ubuntu.com/legal/contributors/agreement), if you are a first time contributor.
* [x] The introduced changes are covered by unit and/or integration tests.

### CODE REVIEW

File: snap/hooks/configure [hunk 1]
```diff
@@ -643,6 +643,7 @@ fi
 if [ -e "${SNAP_DATA}/var/lock/no-flanneld" ]
 then
   touch "${SNAP_DATA}/var/lock/cni-needs-reload"
+  touch "${SNAP_DATA}/var/lock/try-offloading-patch"
```
Comment (neoaggelos) [inline snap/hooks/configure:646]: patch offloading is a one-time configuration, so i'd suggest moving this to `install` instead.

File: microk8s-resources/default-hooks/reconcile.d/10-cni-restart [hunk 2]
```diff
@@ -29,3 +23,12 @@ then
...
   fi
 fi
+
+if  ! [ -e "${SNAP_DATA}/var/lock/skip-offloading" ] &&
+    [ -e "${SNAP_DATA}/var/lock/try-offloading-patch" ] &&
+    ! grep -q "ChecksumOffloadBroken" "${SNAP_DATA}/args/cni-network/cni.yaml" &&
+    "${KUBECTL}" patch felixconfigurations default --patch '{"spec":{"featureDetectOverride":"ChecksumOffloadBroken=true"}}' --type=merge
+then
+    echo "Patched calico for offloading"
+    rm "${SNAP_DATA}/var/lock/try-offloading-patch"
```
Comment (neoaggelos) [inline microk8s-resources/default-hooks/reconcile.d/10-cni-restart:33]: i would suggest `sed s,try-offloading-patch,offloading-patch-applied`

**Review by neoaggelos — COMMENTED** [review]
two small patches, other than that we should be good to go!

**Comment by neoaggelos** [comment]
Since we are on the subject, can you also add the following to the `env` section of `daemonset/calico-node` at `upgrade-scripts/000-switch-to-calico/resources/calico.yaml`?

```
- name: FELIX_FEATUREDETECTOVERRIDE
  value: ChecksumOffloadBroken=true
```

This would effectively prevent the issue from ever manifesting in all new clusters.

See the "manifest" section in https://projectcalico.docs.tigera.io/reference/felix/configuration

File: microk8s-resources/default-hooks/reconcile.d/10-cni-restart [hunk 3]
```diff
@@ -29,3 +23,12 @@ then
     rm "${SNAP_DATA}/var/lock/cni-needs-reload"
   fi
 fi
+
+if  ! [ -e "${SNAP_DATA}/var/lock/skip-offloading" ] &&
+    [ -e "${SNAP_DATA}/var/lock/try-offloading-patch" ] &&
```
Comment (neoaggelos) [inline microk8s-resources/default-hooks/reconcile.d/10-cni-restart:28]: How about this?
```suggestion
    ! [ -e "${SNAP_DATA}/var/lock/offloading-patch-applied" ] &&
```

Then you can also skip the `try-offloading-patch` entirely.

**Review by neoaggelos — CHANGES_REQUESTED** [review]
Almost there, one final fix and it should be good to merge!

File: microk8s-resources/default-hooks/reconcile.d/10-cni-restart [hunk 4]
```diff
@@ -29,3 +23,12 @@ then
     rm "${SNAP_DATA}/var/lock/cni-needs-reload"
   fi
 fi
+
+if  ! [ -e "${SNAP_DATA}/var/lock/skip-offloading" ] &&
+    ! [ -e "${SNAP_DATA}/var/lock/offloading-patch-applied" ] &&
+    ! grep -q "ChecksumOffloadBroken" "${SNAP_DATA}/args/cni-network/cni.yaml" &&
```
[the remark this answers is not in the document]
Comment (berkayoz) [reply microk8s-resources/default-hooks/reconcile.d/10-cni-restart]: Thanks for the reviews, added the required checks.

**Review by neoaggelos — APPROVED** [review]
LGTM, thanks @berkayoz!

**Pull request merged**